sys/exception.c

```diff
@@ -9,4 +9,12 @@ void exc_before_leave(exc_frame_t *kframe) {
 
   if (td->td_flags & TDF_NEEDSWITCH)
     sched_switch(NULL);
+
+  /* First thing after switching to a thread: Process pending signals. */
+  mtx_lock(&td->td_lock);
```

Comment:
Actually, `signotify` barely [marks a thread](http://bxr.su/FreeBSD/sys/kern/kern_sig.c#575) that it needs signal processing (by setting `TDF_NEEDSIGCHK` flag). Actual signal delivery processing happens within [`ast()`](http://bxr.su/FreeBSD/sys/kern/subr_trap.c#270), which is called by `MipsUserIntr` [here](http://bxr.su/FreeBSD/sys/mips/mips/exception.S#774) via [`DO_AST` macro](http://bxr.su/FreeBSD/sys/mips/include/pcb.h#87). Clearly, our equivalent of `ast` is `exc_before_leave`.